libs/langchain/langchain/chains/combine_documents/map_reduce.py

```diff
@@ -2,18 +2,266 @@
+            documents. "page_content" will automatically retrieve the
+            `Document.page_content`, and all other inputs variables will be
+            automatically retrieved from the `Document.metadata` dictionary. Default to
+            a prompt that only contains `Document.page_content`.
+
+    Returns:
+        An LCEL `Runnable` chain.
+
+        Expects a dictionary as input. Input must contain "context" key with a list of
+        Documents.
+
+        Returns a list of Documents, with the contents of each document being the output
+        of passing the corresponding input document to the model. Document order is
+        preserved.
+
+    Example:
+        .. code-block:: python
+
+            from langchain_community.chat_models import ChatOpenAI
+            from langchain_core.documents import Document
+            from langchain_core.prompts import ChatPromptTemplate
+            from langchain.chains.combine_documents import create_map_documents_chain
+
+            llm = ChatOpenAI(model="gpt-3.5-turbo")
+            extract_prompt = ChatPromptTemplate.from_template(
+                [
+                    ("system", "Given a user question, extract the most relevant parts of the following context:\n\n{context}"),
+                    ("human", "{question}"),
+                ]
+            )
+            map_documents_chain = create_map_documents_chain(llm, extract_prompt)
+
+            docs = [
+                Document(page_content="Jesse loves red but not yellow"),
+                Document(page_content = "Jamal loves green but not as much as he loves orange")
+            ]
+
+            map_documents_chain.invoke({"context": docs, "question": "Who loves green?"})
+    """  # noqa: E501
+    _document_prompt = document_prompt or PromptTemplate.from_template("{page_content}")
+
+    def _format_document(inputs: dict) -> str:
+        return format_document(inputs[DOCUMENTS_KEY], _document_prompt)
+
+    map_content_chain = (
+        RunnablePassthrough.assign(**{DOCUMENTS_KEY: _format_document}).with_config(
+            run_name="format_document"
+        )
+        | prompt
+        | llm
+        | StrOutputParser()
+    ).with_config(run_name="map_content")
+
+    map_doc_chain = (
+        RunnablePassthrough.assign(page_content=map_content_chain).with_config(
+            run_name="return_mapped_content"
+        )
+        | RunnableLambda(
+            lambda x: Document(
+                page_content=x["page_content"], metadata=x[DOCUMENTS_KEY].metadata
+            )
+        ).with_config(run_name="compile_document")
+    ).with_config(run_name="map_document")
+
+    def list_inputs(inputs: dict) -> list:
+        docs = inputs.pop(DOCUMENTS_KEY)
+        return [{DOCUMENTS_KEY: doc, **inputs} for doc in docs]
+
+    return (list_inputs | (map_doc_chain.map())).with_config(
+        run_name="map_documents_chain"
+    )
+
+
+def create_collapse_documents_chain(
```

Comment:
is this just reduce? shouldnt this go in reduce file?

Reply:
no strong opinions on where stuff should live, assume everything will be imported from `langchain.chains.combine_documents_chains` anyways